### DISCUSSION TITLE
[A-L] (23 ottobre 2023) Esercizio 3

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#81] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/d012ba71-00d3-4e72-a85c-f5bc58d96216)

**Comment by Frisbon** [comment]
Non ho verificato se funziona ma dovrebbe...
_(sono le 9 di sera e sono >8 ore che faccio algo2)_
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/7593c9bc-0f87-422c-b0f8-feffa69b3dc3)

Il costo è $O(n \cdot (n+m))$ che è **più efficace** della complessità richiesta.
_(dice GPT, io dopo titti me so scordato come si valuta)_
``` python
def esercizio(G):
    cycles = {}
    for x in V(G):
        cycles.add(DFS_getCycleAndCOST(x))
    #ora avrò cycles = { ([sequenza ciclo1],costo ciclo1), ecc. }
    m = (None, INF)
    for t in cycles: #cerca tupla con costo minimo
        if m[1] > t[1]: m = t 
    return m
         
def DFS_getCycleAndCOST(x):
    VIS=[0,...,0]; VIS[x] = 1;
    S = stack(x)
    L = list(x)
    while S is not empty:
        y = S.top()
        if y.out is not empty:
            z = y.out[0]
            if VIS[z] == 0:
                VIS[z] = 1
                S.push(z)
            else:
                cost = peso(y,z)
                counter = 0
                L.add(z)
                while S.top() != z:
                    nodo = S.pop()
                    counter += 1
                    cost += peso(nodo,S.top())
                return (cost, L[-counter])
        else:
            S.pop()
            L.remove(-1)
            cost -= peso(y,z)
    return INF #se non trovo ciclo
```

![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/a7fd76e3-cee2-4f33-aece-309fe40b7866)

**Comment by cornflexxx** [comment]
Come suggerisce il costo computazionale $O(m(n+m)\log n)$ nella soluzione dobbiamo applicare $Dijkstra$ a tutti gli archi di $G$. Prendiamo un qualunque arco $(x,y)\in E(G)$  ed effettuiamo una ricerca dei cammini minimi con $Dijkstra$ a partire da $x$ su $E(G)-(x,y)$. Se l'algoritmo raggiunge comunque $y$ vuol dire che è presente un ciclo $C_x$ nel grafo, e  il peso del ciclo sarà:

$$w(C_x) =w( P[y],y) + w(P[P[y]],P[y]) + ....+ (x,y)$$
Per ottenere il ciclo di costo minimo è sufficiente iterare su tutti gli archi e mantenere memorizzato il ciclo di costo minimo.

**Comment by 18xChris** [comment]
Rientra nella complessita', va bene cosi'
```Python
def find_cycle(G,v,cicli,vis = [0..0]):
  if vis[v] == 1:
    cicli[v] = 1
  vis[v] == 1
  for vv vicino di v:
    return find_cycle(G,vv,cicli,vis)



def solve(G):
  cicli = [0...0]
  for e in E(G):
    find_cyle(G,e.vertice_di_partenza,cicli)

  minimo = inf
  for e in E(G):
    if cicli[e] == 1:
      minimo = min(minimo,dikstra(G,e.vertice_di_arrivo,e.vertice_partenza))

  return minimo```

**Comment by IonutCicio** [comment]
Per ogni arco $(x, y) \in E(G)$, applico dijkstra sul grafo $G$ \ $\set{(x, y)\}$ (partendo da $y$). Se $x$ non è raggiungibile, DIST[x] = ∞, quindi non c'è un ciclo. Se invece $x$ è raggiungibile, allora c'è un ciclo, e il suo costo è w((x, y)) + DIST[x]. A questo punto basta trovare il minimo di questi costi.

```python
min_cycle(G: graph)
    min = ∞
    for (x, y) in E(G)
        DIST = dijkstra(G \ (x, y), y)
        min = min(min, DIST[x] + w((x, y)))
    return min
```

Il costo di dijkstra ottimizzato è $O((n + m) log(n))$, e viene eseguito per $m$ iterazioni del `for`.

**Comment by daromacarlo** [comment]
Strategia greedy di costo O(m(n+m))
Inserisco nella soluzione gli archi dal più leggero al più pesante, appena trovo un ciclo, questo sarà il più leggero, lo cerco con la funzione CercaCiclo e lo ritorno
![Immagine 2025-01-28 170025](https://github.com/user-attachments/assets/569b6bab-f755-4699-b365-3ecaadd271c1)
![Immagine 2025-01-28 170037](https://github.com/user-attachments/assets/3626361a-f441-4e7d-86f2-dd71b14356f7)
![Immagine 2025-01-28 170049](https://github.com/user-attachments/assets/e4878ac9-d233-4065-9463-dbdab834d659)
![Immagine 2025-01-28 170058](https://github.com/user-attachments/assets/d3fd39bd-2407-4761-a4d7-8a2c9a2d869f)
